Update Zig from 0.10.1 to 0.11.0

kprotty opened the pull request:
With this, we should finally be tracking the latest tagged release. View the [release notes](https://ziglang.org/download/0.11.0/release-notes.html) for some relevant information on upgrading.

kprotty committed:
zig: wip builtin changes

kprotty committed:
zig: update math.min/max to builtins

kprotty committed:
zig: update `.field_type` from Type structs to `.type`

kprotty committed:
zig: build.zig updates

kprotty committed:
zig: alignForward and bitCount updates

kprotty committed:
zig: remove hash_map.zig for std.hash_map

The backported fix is available in 0.11

kprotty committed:
zig: update allocAdvanced to alignedAlloc

kprotty committed:
zig: update mem.set to `@memset` builtin

kprotty committed:
zig: remove union_field_parent_ptr for fieldParentPtr

kprotty committed:
zig: various fixes

kprotty committed:
zig: build/benchmark/lint work

kprotty committed:
zig; build test works

kprotty committed:
zig: more fixes for clients

kprotty committed:
Merge remote-tracking branch 'origin/main' into king-upgrade-zig11

kprotty committed:
zig: merge fixes

matklad commented on src/aof.zig:
```diff
@@ -59,13 +59,13 @@ pub const AOFEntry = extern struct {
 
     pub fn header(self: *AOFEntry) *Header {
-        return @ptrCast(*Header, &self.message);
+        return @as(*Header, @ptrCast(&self.message));
```
After this PR is merger, we need to do a pass over the codebase to check how much of these `@as` could be elliminated. No need to do this in this PR though!

matklad commented on src/hash_map.zig:
```diff
@@ -104,7 +104,7 @@ pub const StringIndexContext = struct {
 
     pub fn hash(self: @This(), x: u32) u64 {
-        const x_slice = mem.sliceTo(@ptrCast([*:0]const u8, self.bytes.items.ptr) + x, 0);
+        const x_slice = mem.sliceTo(@as([*:0]const u8, @ptrCast(self.bytes.items.ptr)) + x, 0);
```
for hash map, I think we can now just switch to upstream impl?

matklad replied:
I guess I'll see this in a further commit!

matklad commented on src/testing/storage.zig:
```diff
@@ -506,19 +506,13 @@ pub const Storage = struct {
-            @alignOf(superblock.SuperBlockHeader),
-            mem.bytesAsValue(superblock.SuperBlockHeader, bytes),
-        );
+        return @alignCast(mem.bytesAsValue(superblock.SuperBlockHeader, bytes));
```
Oh, I think I need to check if we now can introduce `stdx.bytes_as_value_but_give_error_at_compile_time_if_I_mess_alignment_please`!

matklad commented on src/constants.zig:
```diff
@@ -298,7 +298,7 @@ pub const connection_delay_min_ms = config.process.connection_delay_min_ms;
 
 /// The maximum number of outgoing messages that may be queued on a replica connection.
-pub const connection_send_queue_max_replica = std.math.max(std.math.min(clients_max, 4), 2);
+pub const connection_send_queue_max_replica = @max(@min(clients_max, 4), 2);
```
After this PR we should add `stdx.clamp`, the minmax combo is too easy to mess up,

matklad commented on src/vsr/replica.zig:
```diff
@@ -8569,10 +8569,10 @@ const DVCQuorum = struct {
-            commit_max_ = std.math.max(commit_max_, dvc_commit_max_tail);
-            commit_max_ = std.math.max(commit_max_, dvc_commit_max_pipeline);
-            commit_max_ = std.math.max(commit_max_, dvc.header.commit);
-            commit_max_ = std.math.max(commit_max_, dvc_headers.slice[0].commit);
```
TODO for after PR: this would be _perfect_ to use vararg capability of `@max`

matklad commented on src/config.zig:
```diff
@@ -168,7 +168,7 @@ const ConfigCluster = struct {
     pub fn checksum(comptime config: ConfigCluster) u128 {
         @setEvalBranchQuota(10_000);
-        var hasher = std.crypto.hash.Blake3.init(.{});
+        var hasher = std.crypto.hash.blake2.Blake2s256.init(.{});
```
Is Blake3 == Blake2s256? If it is, might make sense to add a comment here. Also, maybe `TODO: switch to vsr.checksum once that works at comptime`? (I tihnk I have an issue tracking that,but comment won't hurt)

kprotty replied:
Good point on the comment. It was changed due to Blake3 in stdlib having a compile error internally.

matklad commented on src/vsr/superblock_quorums.zig:
```diff
@@ -97,7 +97,7 @@ pub fn QuorumsType(comptime options: Options) type {
 
             for (copies, 0..) |*copy, index| quorums.count_copy(copy, index, threshold);
 
-            std.sort.sort(Quorum, quorums.slice(), {}, sort_priority_descending);
```
Unreleated, but I think `sort.insertion` actually would be good here, as the size is small, and we could save on codegen here. I wonder if should add `stdx.tiny_sorts` which asserts that the size of the slice is `<= 32`?

matklad reviewed (commented):
Firs batch it!

kassane commented:
Is it intended add **zig-pkg** support for `c_client`?

Missing `lib.installHeader("Path/tb_client.h", "tb_client.h");` copying the Tigerbeetle package header to the client application.

### My next update to tb-cpp using zig-pkg

#### My `build.zig` app:

```zig
const libtb_dep = b.dependency("tigerbeetle", .{
        .target = info.target,
        .optimize = info.optimize,
    });
const libtb = libtb_dep.artifact("tigerbeetle");


// get includepath from pkg to zig-cache/i/{hash-pkg}/include
for (libtb.include_dirs.items) |include| {
            exe.include_dirs.append(include) catch @panic("Unavailable include_dirs");
        }

// or to zig-out/include
lib.installLibraryHeaders(libtb);

exe.linkLibrary(libtb);
```

matklad commented:
>Is it intended add zig-pkg support for c_client?

Probably not right off the bat, as the immediate goal is to just move the codebase over to the new zig. 

Once we are over that hurdle though, we could indeed extract relevant bits of TigerBeetle into proper zig packages.

matklad commented on build.zig:
```diff
@@ -1149,9 +1250,9 @@ const GitCloneStep = struct {
 /// - From MacOS, similar to https://github.com/ziglang/zig/issues/9711#issuecomment-1090071087.
 /// Workaround: Just setting different cache folders for each platform.
 fn builder_with_isolated_cache(
```
After the PR lands, we need to check if we can remove this altogether, perhaps the underlying bugs were fixed?

matklad commented on src/stdx.zig:
```diff
@@ -7,8 +7,8 @@ const assert = std.debug.assert;
-pub const HashMapUnmanaged = @import("./hash_map.zig").HashMapUnmanaged;
-pub const AutoHashMapUnmanaged = @import("./hash_map.zig").AutoHashMapUnmanaged;
+pub const HashMapUnmanaged = std.HashMapUnmanaged;
+pub const AutoHashMapUnmanaged = std.AutoHashMapUnmanaged;
```
:tada:

matklad commented on src/lsm/set_associative_cache.zig:
```diff
@@ -570,11 +569,11 @@ fn PackedUnsignedIntegerArray(comptime UInt: type) type {
+    assert(@typeInfo(UInt).Int.bits < @bitSizeOf(u8));
     assert(math.isPowerOfTwo(@typeInfo(UInt).Int.bits));
 
-    const word_bits = @bitCount(Word);
```
Wait, what happened here? `@bitCount` isn't a Zig builtin... Is this some ancient dead code?

kprotty replied:
It was `std.meta.bitCount` which I did a find-and-replace to `@bitCount` then realized "wait, that doesn't exist" and fixed it back to `@bitSizeOf`.

matklad commented on src/stdx.zig:
```diff
@@ -428,50 +428,3 @@ pub fn update(base: anytype, diff: anytype) @TypeOf(base) {
-// TODO(Zig): No need for this function once Zig is upgraded
-// and @fieldParentPtr() can be used for unions.
-// See: https://github.com/ziglang/zig/issues/6611.
-pub inline fn union_field_parent_ptr(
```
:tada: :tada: :tada:

matklad commented on src/aof.zig:
```diff
@@ -431,7 +431,7 @@ pub fn aof_merge(
     try stdout.print("Building checksum map...\n", .{});
     var current_parent: ?u128 = null;
-    for (aofs[0..aof_count], 0..) |*aof, i| {
+    for (&aofs[0..aof_count], 0..) |*aof, i| {
```
TODO: we also probably want convert a bunch of while loops to `for (0..n) |i|` form...

matklad commented on src/vsr/replica.zig:
```diff
@@ -3893,7 +3893,7 @@ pub fn ReplicaType(
 
             message.header.* = .{
-                .size = @as(u32, @intCast(@sizeOf(Header) * (1 + self.view_headers.array.len))),
+                .size = @as(u32, @intCast(@as(usize, @sizeOf(Header)) * (1 + self.view_headers.array.len))),
```
I wonder we could reduce this to a signle `@as`

matklad commented on src/clients/dotnet/dotnet_bindings.zig:
```diff
@@ -134,23 +134,21 @@ fn get_mapped_type_name(comptime Type: type) ?[]const u8 {
 }
 
 fn to_case(comptime input: []const u8, comptime case: enum { camel, pascal }) []const u8 {
-    comptime {
```
Why do we need to return `comptime` here? Seems important to signal that this code is intended to be run at comptime.

kprotty replied:
`comptme blk:` has the same effect (the code in the block runs at comptime), but doesn't hit the issue of the compiler thinking `return output[0..len]` happens at runtime (even inside `comptime {}`) for some reason.

matklad commented on src/clients/go/go_bindings.zig:
```diff
@@ -46,7 +46,9 @@ fn get_mapped_type_name(comptime Type: type) ?[]const u8 {
 
 fn to_pascal_case(comptime input: []const u8, comptime min_len: ?usize) []const u8 {
-    comptime {
+    // TODO: Cleanup when this is fixed after Zig 0.11.
```
Let's use more greppable `TODO(Zig)`

matklad commented on src/flags.zig:
```diff
@@ -489,7 +489,13 @@ test "flags" {
+                error.EnvironmentVariableNotFound => return error.SkipZigTest,
+                error.InvalidUtf8 => unreachable,
+                error.OutOfMemory => return error.OutOfMemory,
+            };
```
`defer gpa.free(zig_exe)`

matklad commented on src/testing/cluster.zig:
```diff
@@ -42,7 +42,9 @@ pub const Failure = enum(u8) {
 const client_id_permutation_shift = constants.members_max;
 
-pub fn ClusterType(comptime StateMachineType: fn (comptime Storage: type, comptime constants: anytype) type) type {
+// TODO: Once Zig is upgraded from 0.11, change StateMachineType from anytype back to
```
`TODO(Zig)`

matklad commented on src/clients/go/docs.zig:
```diff
@@ -11,11 +12,20 @@ fn go_current_commit_pre_install_hook(
             .{ sample_dir, file },
-        )) catch {
+        );
+        // TODO: Mismatching internal std.os.unlink() error in Zig 0.11.0
```
TODO(Zig)

kprotty committed:
zig: fixing review comments

kprotty committed:
zig: `-Drelease-safe` to `-Doptimize=ReleaseSafe`

kprotty committed:
Merge remote-tracking branch 'origin/main' into king-upgrade-zig11

kprotty committed:
build: fix linux code path

matklad commented:
Ok, this looks good to me! Compiled post-merge checklist from non-blocking comments here: 

* https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1291758901 needless `@as` cast
* https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1291766926 add alignment-preseving versions of asBytes
* https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1291773386 stdx.clamp.
* https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1291774463 `@max` is now vararg
* !!! https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1291794038 audit what's up with our config checksum. I _think_ we leak that into various places on disk
* https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1291797718 stdx.tiny_sort
* https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1293390649 remove builder_works_with_isolated_cache
* https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1293409722 use `for(0..n) |i|`
* https://github.com/tigerbeetle/tigerbeetle/pull/1112#discussion_r1293412872 more `@as`

Client builds seem to fail though.

kprotty commented:
For the client builds, seems like an error in build.zig when porting `setOutputDir()` as the binaries aren't going to their correct folders.

batiati commented on src/lsm/manifest_level.zig:
```diff
@@ -546,7 +546,7 @@ pub fn ManifestLevelType(
                 if (optimal == null or range.tables.len < optimal.?.range.tables.len) {
                     optimal = LeastOverlapTable{
                         .table = TableInfoReference{
-                            .table_info = @intToPtr(*TableInfo, @ptrToInt(table)),
```
We can replace this pattern `@intToPtr(*X, @ptrToInt(x_const))` with `@constCast` introduced in Zig 0.11.0.

kprotty committed:
Merge branch 'main' into king-upgrade-zig11

kprotty committed:
build: zig-cache relative setOutputDir

kprotty committed:
test: fix errors detected by `tidy.zig`

- `node_pool.zig` line too long
- `static_allocator.zig` no longer exceeding 100 char col

kprotty committed:
Merge branch 'king-upgrade-zig11' of https://github.com/tigerbeetle/tigerbeetle into king-upgrade-zig11

matklad reviewed (approved):
Approving!

This currently fails client builds, but I think that it would be less pain overall fixing that after the merge, rather than rebasing this continiously (less paint still means a lot of pain though!)
